File: interpreter_node/src/rg/transformators/skipSelfAssignments.ts
import * as ast from '../ast';

export function skipSelfAssignments({ edges }: ast.GameDeclaration) {
  for (const edge of edges) {
    if (isSelfAssignment(edge.label)) {
      edge.label = ast.Skip({});
    }
  }
}

function isEqualReference(lhs: ast.Expression, rhs: ast.Expression): boolean {
  if (lhs.kind === 'Cast') {
    return isEqualReference(lhs.rhs, rhs);
  }

  if (rhs.kind === 'Cast') {
    return isEqualReference(lhs, rhs.rhs);
  }

  switch (lhs.kind) {
    case 'Access':
      return (
        rhs.kind === 'Access' &&
        isEqualReference(lhs.lhs, rhs.lhs) &&
        isEqualReference(lhs.rhs, rhs.rhs)
      );
    case 'Reference':
      return rhs.kind === 'Reference' && lhs.identifier === rhs.identifier;
  }
}

function isSelfAssignment(edgeLabel: ast.EdgeLabel) {
  return (
    edgeLabel.kind === 'Assignment' &&
    isEqualReference(edgeLabel.lhs, edgeLabel.rhs)
  );
}

File: interpreter_rust/rg_transform/src/skip_self_assignments.rs
use rg::ast::{EdgeDeclaration, EdgeLabel, Error, GameDeclaration};
use std::rc::Rc;

pub fn skip_self_assignments<Id: PartialEq>(
    mut game_declaration: GameDeclaration<Id>,
) -> Result<GameDeclaration<Id>, Error<Id>> {
    for edge in &mut game_declaration.edges {
        if edge.label.is_self_assignment() {
            *edge = Rc::new(EdgeDeclaration {
                label: Rc::new(EdgeLabel::Skip),
                lhs: edge.lhs.clone(),
                rhs: edge.rhs.clone(),
            });
        }
    }

    Ok(game_declaration)
}
